[MPS] Add `aten::angle`

sezelt opened the pull request:
This adds an MPS backend implementation for `aten::angle` and `aten::angle_out` (mentioned in issue #77764), following the example #78408.

sezelt committed:
add MPS support for angle

sezelt committed:
run lintrunner

malfet commented:
Looks good to me in theory, but it needs a special case for handling non-complex types, as documented in https://pytorch.org/docs/stable/generated/torch.angle.html

sezelt committed:
test for propagation of NaN in MPS aten::angle

sezelt committed:
move MPS aten::angle into UnaryOps.mm and correct choice of output dtype to match CPU

sezelt committed:
add aten::angle to supported complex ops list in test_mps.py

sezelt committed:
check for complex support in MPS aten::angle

sezelt committed:
create special case for NaN in MPS aten::angle

sezelt committed:
run lintrunner

sezelt committed:
fix bug in NaN propagation in MPS aten::angle

sezelt committed:
mark TestMPS.test_angle to expect failure on macOS < 14

sezelt commented:
For real inputs, `imaginaryPartOfTensor` returns zeroes so the behavior is correct there, I added `float32` input to the tests to test to make sure. I've updated it so that it returns `float32` for integer inputs to match CPU behavior. I also moved the function into `UnaryOps.mm` and used the template in there rather than have a separate file. I updated the pass/fail expectations in `TestConsistency` and the consistency checks are now passing. On macOS 13 it's limited to only real inputs but does behave correctly for them.

malfet commented:
@pytorchbot merge -f "Lint + MPS tests are green"
